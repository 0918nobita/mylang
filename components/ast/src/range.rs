use std::fmt;

use serde::{Deserialize, Serialize};

use super::pos::Pos;

#[derive(Clone, Default, Serialize, Deserialize)]
pub struct Range {
    pub start: Pos,
    pub end: Pos,
}

impl Range {
    pub fn new(start: Pos, end: Pos) -> Self {
        Range { start, end }
    }
}

impl fmt::Debug for Range {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}-{:?}", self.start, self.end)
    }
}

impl fmt::Display for Range {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}-{:?}", self.start, self.end)
    }
}

impl From<Pos> for Range {
    fn from(pos: Pos) -> Self {
        Self {
            start: pos.clone(),
            end: pos,
        }
    }
}

pub trait Locatable {
    fn locate(&self) -> Range;
}
