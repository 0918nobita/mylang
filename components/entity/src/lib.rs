use std::fmt::{Debug, Formatter, Result as FmtResult};

#[derive(PartialEq)]
pub struct I32Entity(i32);

impl I32Entity {
    pub fn new(value: i32) -> Self {
        Self(value)
    }

    pub fn add(&self, rhs: &Self) -> Self {
        Self(self.0 + rhs.0)
    }
}

impl Debug for I32Entity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

#[derive(PartialEq)]
pub struct StrEntity(String);

impl StrEntity {
    pub fn new(value: String) -> Self {
        Self(value)
    }
}

impl Debug for StrEntity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

#[derive(Debug)]
pub enum Entity {
    I32(I32Entity),
    Str(StrEntity),
}
