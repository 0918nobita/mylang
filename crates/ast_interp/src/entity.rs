//! 抽象構文木インタプリタで扱う「実体」の定義

use std::fmt::{Debug, Display, Formatter, Result as FmtResult};

/// 符号付き32ビット整数値
pub struct I32Entity(i32);

impl I32Entity {
    pub fn new(value: i32) -> Self {
        Self(value)
    }

    pub fn add(&self, rhs: &Self) -> Self {
        Self(self.0 + rhs.0)
    }
}

impl Debug for I32Entity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

impl Display for I32Entity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

/// 文字列データ
pub struct StrEntity(String);

impl StrEntity {
    pub fn new(value: String) -> Self {
        Self(value)
    }
}

impl Debug for StrEntity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

impl Display for StrEntity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

/// 実行時型情報
#[derive(Debug)]
pub enum RuntimeTypeInfo {
    I32,
    Str,
}

impl Display for RuntimeTypeInfo {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(
            f,
            "{}",
            match self {
                RuntimeTypeInfo::I32 => "i32",
                RuntimeTypeInfo::Str => "str",
            }
        )
    }
}

/// バイトコード実行時に扱う値
#[derive(Debug)]
pub enum Entity {
    I32(I32Entity),
    Str(StrEntity),
}

impl Entity {
    pub fn get_type(&self) -> RuntimeTypeInfo {
        match self {
            Entity::I32(_) => RuntimeTypeInfo::I32,
            Entity::Str(_) => RuntimeTypeInfo::Str,
        }
    }
}
