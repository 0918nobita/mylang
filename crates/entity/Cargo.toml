[package]
name = "mylang_entity"
version = "0.1.0"
edition = "2021"

[dependencies]
