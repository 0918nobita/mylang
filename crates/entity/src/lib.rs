use std::fmt::{Debug, Display, Formatter, Result as FmtResult};

use serde::Serialize;

/// 符号付き32ビット整数値
#[derive(PartialEq, Serialize)]
pub struct I32Entity(i32);

impl I32Entity {
    pub fn new(value: i32) -> Self {
        Self(value)
    }

    pub fn add(&self, rhs: &Self) -> Self {
        Self(self.0 + rhs.0)
    }
}

impl Debug for I32Entity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

impl Display for I32Entity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

/// 文字列データ
#[derive(PartialEq, Serialize)]
pub struct StrEntity(String);

impl StrEntity {
    pub fn new(value: String) -> Self {
        Self(value)
    }
}

impl Debug for StrEntity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

impl Display for StrEntity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

/// バイトコード実行時に扱う値
#[derive(Debug, Serialize)]
pub enum Entity {
    I32(I32Entity),
    Str(StrEntity),
}
