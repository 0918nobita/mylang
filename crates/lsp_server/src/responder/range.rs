use mylang_token::Locatable;
use serde_json::Value as JsonValue;

pub fn locatable_to_lsp_range<L>(locatable: &L) -> JsonValue
where
    L: Locatable,
{
    let range = locatable.locate();
    serde_json::to_value(range).expect("Failed to convert locatable to json")
}
