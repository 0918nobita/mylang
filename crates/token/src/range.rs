use std::fmt;

use serde::{Deserialize, Serialize};

use super::pos::Pos;

/// ソースコード中の範囲
#[derive(Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct Range {
    /// 始端
    start: Pos,
    /// 終端
    end: Pos,
}

impl Range {
    /// 新しい範囲を生成して返す
    pub fn new(start: Pos, end: Pos) -> Self {
        Range { start, end }
    }

    /// 終端の参照を返す
    pub fn end_ref(&self) -> &Pos {
        &self.end
    }

    /// 所有権をムーブして終端を返す
    pub fn end(self) -> Pos {
        self.end
    }

    /// 終端を別の範囲の終端に設定する
    ///
    /// ```
    /// use token::range;
    ///
    /// let mut former = range!(0;0, 0;5);
    ///
    /// let latter = range!(0;7, 0;10);
    /// former.concat(latter);
    ///
    /// let expected = range!(0;0, 0;10);
    /// assert_eq!(former, expected);
    /// ```
    pub fn concat(&mut self, other: Range) {
        self.end = other.end;
    }
}

impl fmt::Debug for Range {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}-{:?}", self.start, self.end)
    }
}

impl fmt::Display for Range {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}-{}", self.start, self.end)
    }
}

impl From<Pos> for Range {
    /// 始端と終端が同じ位置の範囲に変換する
    fn from(pos: Pos) -> Self {
        Self {
            start: pos.clone(),
            end: pos,
        }
    }
}
