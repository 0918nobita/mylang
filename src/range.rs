use std::fmt;

use super::pos::Pos;

#[derive(Clone, Default)]
pub struct Range {
    start: Pos,
    end: Pos,
}

impl fmt::Debug for Range {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}-{:?}", self.start, self.end)
    }
}

pub trait Locatable {
    fn get_range(&self) -> Option<Range>;
}
