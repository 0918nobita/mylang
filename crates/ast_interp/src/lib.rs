//! 抽象構文木インタプリタ
//!
//! スモールステップ意味論に沿って、構文木を繰り返し簡約し結果を返す。

mod entity;

use mylang_ast::{Expr, Stmt};
use thiserror::Error;

use entity::{Entity, I32Entity, RuntimeTypeInfo, StrEntity};

#[derive(Debug, Error)]
pub enum AstInterpError {
    #[error("Type mismatch (expected: {expected}, actual: {actual})")]
    TypeMismatch {
        expected: RuntimeTypeInfo,
        actual: RuntimeTypeInfo,
    },
}

pub type AstInterpResult<T> = Result<T, AstInterpError>;

fn eval_expr(expr: &Expr) -> AstInterpResult<Entity> {
    match expr {
        Expr::I32Lit(_, i) => Ok(Entity::I32(I32Entity::new(*i))),

        Expr::StrLit(_, s) => Ok(Entity::Str(StrEntity::new(s.clone()))),

        Expr::Add(lhs, rhs) => {
            let lhs = eval_expr(lhs)?;

            let lhs = match lhs {
                Entity::I32(i32_entity) => i32_entity,
                _ => {
                    return Err(AstInterpError::TypeMismatch {
                        expected: RuntimeTypeInfo::I32,
                        actual: lhs.get_type(),
                    })
                }
            };

            let rhs = eval_expr(rhs)?;

            let rhs = match rhs {
                Entity::I32(i32_entity) => i32_entity,
                _ => {
                    return Err(AstInterpError::TypeMismatch {
                        expected: RuntimeTypeInfo::I32,
                        actual: rhs.get_type(),
                    })
                }
            };

            Ok(Entity::I32(lhs.add(&rhs)))
        }
    }
}

/// 抽象構文木を解釈実行する
pub fn execute(stmts: &[Stmt]) -> AstInterpResult<()> {
    for stmt in stmts {
        match stmt {
            Stmt::PrintI32(_, expr) => {
                let val = eval_expr(expr)?;

                let val = match val {
                    Entity::I32(i32_entity) => i32_entity,
                    _ => {
                        return Err(AstInterpError::TypeMismatch {
                            expected: RuntimeTypeInfo::I32,
                            actual: val.get_type(),
                        })
                    }
                };

                println!("{}", val);
            }
            Stmt::PrintStr(_, expr) => {
                let val = eval_expr(expr)?;

                let val = match val {
                    Entity::Str(str_entity) => str_entity,
                    _ => {
                        return Err(AstInterpError::TypeMismatch {
                            expected: RuntimeTypeInfo::Str,
                            actual: val.get_type(),
                        })
                    }
                };

                println!("{}", val);
            }
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use mylang_ast::Expr;
    use mylang_token::range;

    use super::eval_expr;

    #[test]
    /// 整数リテラルの評価結果が変化していないことを確認する
    fn i32_lit() {
        let ent = eval_expr(&Expr::I32Lit(range!(0;0), 2));
        insta::assert_debug_snapshot!(ent);
    }

    #[test]
    /// 整数リテラル同士の足し算の評価結果が変化していないことを確認する
    fn i32_lit_plus_i32_lit() {
        let ent = eval_expr(&Expr::Add(
            Box::new(Expr::I32Lit(range!(0;0), 2)),
            Box::new(Expr::I32Lit(range!(0;0), 3)),
        ));
        insta::assert_debug_snapshot!(ent);
    }

    #[test]
    /// 文字列リテラルの評価結果が変化していないことを確認する
    fn str_lit() {
        let ent = eval_expr(&Expr::StrLit(range!(0;0), "foo".to_owned()));
        insta::assert_debug_snapshot!(ent);
    }

    #[test]
    /// 整数リテラルと文字列リテラルの足し算で発生するエラーの内容が変化していないことを確認する
    fn i32_lit_plus_str_lit() {
        let res = eval_expr(&Expr::Add(
            Box::new(Expr::I32Lit(range!(0;0), 3)),
            Box::new(Expr::StrLit(range!(0;0), "bar".to_owned())),
        ));
        insta::assert_debug_snapshot!(res);
    }

    #[test]
    /// 文字列リテラルと整数リテラルの足し算で発生するエラーの内容が変化していないことを確認する
    fn str_lit_plus_i32_lit() {
        let res = eval_expr(&Expr::Add(
            Box::new(Expr::StrLit(range!(0;0), "foo".to_owned())),
            Box::new(Expr::I32Lit(range!(0;0), 4)),
        ));
        insta::assert_debug_snapshot!(res);
    }

    #[test]
    /// 文字列リテラル同士の足し算で発生するエラーの内容が変化していないことを確認する
    fn str_lit_plus_str_lit() {
        let res = eval_expr(&Expr::Add(
            Box::new(Expr::StrLit(range!(0;0), "foo".to_owned())),
            Box::new(Expr::StrLit(range!(0;0), "bar".to_owned())),
        ));
        insta::assert_debug_snapshot!(res);
    }
}
