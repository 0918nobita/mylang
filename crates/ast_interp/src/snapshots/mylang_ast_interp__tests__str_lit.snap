---
source: crates/ast_interp/src/lib.rs
assertion_line: 127
expression: ent

---
Ok(
    Str(
        foo,
    ),
)
